sys/netbsd/lwp.txt

```diff
@@ -0,0 +1,33 @@
+# Copyright 2019 syzkaller project authors. All rights reserved.
+# Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+include <sys/types.h>
+include <sys/lwp.h>
+include <sys/lwpctl.h>
+
+# _lwp_create(context, flags intptr, lwpid ptr [in, int32])
```

Comment:
That one is actually the most important one as it creates more threads.

Reply:
Yeah. I am trying to make the headerparser utility work so that I can parse the necessary structures.

Reply:
Please check for example usage in /usr/src/tests/lib/libc/sys/t_ptrace_wait.c

The most basic usage is along these lines:

```
static void
lwp_main_func(void *arg)
{
        the_lwp_id = _lwp_self();
        _lwp_exit();
}

{
        ucontext_t uc;
        lwpid_t lid;
        static const size_t ssize = 16*1024;
        void *stack;
       stack = malloc(ssize);
_lwp_makecontext(&uc, lwp_main_func, NULL, NULL, stack, ssize); 
_lwp_wait(lid, NULL)
}
```

Reply:
hint can be used for implementing mutexes.

https://nxr.netbsd.org/xref/src/libexec/ld.elf_so/rtld.c#_rtld_mutex

Just grep for its usage.

One thing that we shall be aware of is that we can accidentally deadlock the process and hopefully it won't cause a false positive in syzkaller.

Reply:
Having issues with retrieving details about the __context_t structure. Since the structure's members are other structures in the format

typedef struct {
..
}

Headerparser isn't able to retrieve such a structure. This would require some manual work to get it working.

Reply:
this is part of the reason why C would be much easier here.. but on the other hand it's a benefit to fuzz it with non-conventional input/usage.

Reply:
I have written a script which replaces structure definitions in header files. The parser also doesn't seem to go through variable typedefs. So I needed to add that also. And function definitions in headerfiles were crashing it.

This script does part of the job - I would do the rest manually
https://github.com/R3x/netbsd-fuzzing-aids/blob/master/modify_struct.py

Reply:
I have been able to parse ucontext_t partially with header parser 
`
__ucontext {
          uc_flags        len|fileoff|int32     #(unsigned int)
          uc_link         ucontext_t*           #(ucontext_t*)
          uc_sigmask      sigset_t              #(sigset_t)
          uc_stack        stack_t               #(stack_t)
          uc_mcontext     mcontext_t            #(mcontext_t)
}
mcontext_t {
          __gregs         len|fileoff|flags|intN     #(unsigned long)
          _mc_tlsbase     len|fileoff|flags|intN     #(unsigned long)
          __fpregs        __fpregset_t               #(__fpregset_t)
}
sigset_t {
          __bits     array     #(__uint32_t[4])
}
`
Need to complete stack_t parsing though.